Add events to notify of drawer opening / closing

humancopy opened the pull request:
Hey :)

I think it's a good idea to start using events in Timber instead of callbacks to make it easy to tap into interface changes. I also want to add events to the ajax_cart.js file, what you think?

cshold commented:
I agree this is a great approach. Are you able to update the docs to include these as well?

humancopy commented:
Yes, would be happy to do it. I'll add it to this pull request. :)

humancopy commented:
I added more triggers and fixed the namespace. I've implemented this in a few projects and it's been working great.

Let me know what we can do to push this forward, I'm happy to help with anything needed. 

Have a beautiful weekend.

Peace ;)

Cam commented:
Awesome. This is exactly what I was talking about a little while back. Thanks heaps @humancopy 😃

cshold commented on assets/ajax-cart.js.liquid:
```diff
@@ -72,18 +78,23 @@ ShopifyAPI.onError = function(XMLHttpRequest, textStatus) {
 ==============================================================================*/
 ShopifyAPI.addItemFromForm = function(form, callback, errorCallback) {
-  var params = {
+  var $body = $('body'),
```
A bit of a nitpick, but can you change this to `$(document.body)` - it's slightly faster ([source](https://www.sitepoint.com/jquery-body-on-document-on/)).

humancopy replied:
Done, changed it in a couple of other places also ;)

cshold commented on assets/ajax-cart.js.liquid:
```diff
@@ -39,20 +39,26 @@ ShopifyAPI.onCartUpdate = function(cart) {
     data: 'note=' + attributeToString(note),
     dataType: 'json',
+    beforeSend: function() {
+      $body.trigger('updateCartNote.ajaxCart', note);
```
What's the need for this event, rather than just on success?

humancopy replied:
I changed all to before / after / error / complete events. The before trigger I didn't use but thought to add it as I was already there. :)

cshold commented on assets/ajax-cart.js.liquid:
```diff
@@ -325,7 +357,7 @@ var ajaxCart = (function(module, $) {
   cartCallback = function(cart) {
     $body.removeClass('drawer--is-loading');
-    $body.trigger('ajaxCart.afterCartLoad', cart);
+    $body.trigger('afterCartLoad.ajaxCart', cart);
```
Can you revert this name. It's the only breaking change I can see in the PR

cshold replied:
Scratch that. I'll update the version number differently instead. All good

cshold commented on layout/theme.liquid:
```diff
@@ -424,7 +424,7 @@
       });
 
-      jQuery('body').on('ajaxCart.afterCartLoad', function(evt, cart) {
+      jQuery(document.body).on('ajaxCart.afterCartLoad', function(evt, cart) {
```
Please update these two lines to the new name `afterCartLoad.ajaxCart`

humancopy replied:
Oops, done ;)

cshold commented on assets/ajax-cart.js.liquid:
```diff
@@ -92,40 +106,55 @@ ShopifyAPI.addItemFromForm = function(form, callback, errorCallback) {
     else {
       ShopifyAPI.onCartUpdate(cart);
     }
+    $(document.body).trigger('afterGotCart.ajaxCart', cart);
```
Last change, switch this to `afterGetCart`. Realizing the naming issues as I'm writing up some new docs

humancopy replied:
No worries, done... Let me know if need any help with docs. :)

cshold commented on assets/ajax-cart.js.liquid:
```diff
@@ -354,6 +386,7 @@ var ajaxCart = (function(module, $) {
 
       // If it has a data-line, update the cart.
       // Otherwise, just update the input's number
+      $body.trigger('adjustCart.ajaxCart', [this, line, qty]);
```
Can remove this. `updateQuantity` (the function right after this) calls `ShopifyAPI.changeItem`, which already has triggers fired (`before/after/error/complete ChangeItem`)

humancopy replied:
updateQuantity is not always getting triggered which is why I put this one here

cshold commented on assets/ajax-cart.js.liquid:
```diff
@@ -375,6 +408,7 @@ var ajaxCart = (function(module, $) {
 
       // If it has a data-line, update the cart
       if (line) {
+        $body.trigger('adjustCart.ajaxCart', [this, line, qty]);
```
Remove this line as well, as per the previous comment.

humancopy replied:
Another reason is access to the element which isn't available inside updateQuantity

cshold replied:
The entire cart reprints, so having the element available won't do much help. Open to a rebuttal though.

humancopy replied:
I was using it mainly for when the quantity in the cart was changed and then to easily get the id of the variant but that can easily be done using line for the position of the item in the cart and grabbing it from the input ... So I'll remove this, you're right it's just redundant and confusing... :)

cshold commented:
Alright, we're almost there.

Let's remove:
- `beforeSubmit.ajaxCart` | duplicate of event in `ShopifyAPI.addItemFromForm`
- `afterSubmit.ajaxCart` | duplicate, same as above
- `itemError.ajaxCart` | duplicate, `ShopifyAPI.addItemFromForm` triggers `errorAddItem.ajaxCart` regardless if an error callback is sent
- `adjustCart.ajaxCart` | duplicate of events triggered in `ShopifyAPI.changeItem`
  - My argument with this one is that if `ShopifyAPI.changeItem` isn't called, the events aren't needed at all. Every other event is tied to an ajax call except this one.

humancopy commented:
Removed redundant events. Docs look really good, well done! :)

cshold commented:
Thanks for the updates. Did a bunch of testing and it looks in order. The only duplicate events triggered are `afterCartLoad.ajaxCart` and `afterGetCart.ajaxCart` — but I'm happy to leave both. The second one is in the `ShopifyAPI.getCart` function itself, firing no matter what callback is sent, while the first one is specific to Timber's drawer implementation. 

Will give the docs a quick review the merge both. Cheers to putting this together.

cshold commented:
Last ask, could you squash the commits down to one. [Here are the steps](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/).

humancopy commented:
Unfortunately I can't squash because I merged stuff from the Shopify/Timber as I needed to keep it up to date cause I was using it for some projects, sorry. If you have an idea how to resolve this I'm happy to follow... ;)

cshold commented:
No worries, more of a nice to have than anything.
